### ISSUE TITLE
libtdjson Memory Leak

### ISSUE THREAD

**Issue by anwsonwsymous** [aCLr/rust-tdlib#19]
When I run my application, at the beginning it uses like 18MB, but it increases infinitely. 
In a day it reaches 150MB

I don't know it's `tdlib` bindings issue or `libtdjson `itself. 

I have checked it with `heaptrack` and it shows me that "Leak" comes from `libtdjson` but I am not sure.

Does anyone has idea how to find out what's going on and fix it somehow?
Maybe someone else has the same problem?

**Comment by antonio-antuan** [comment]
I have just searched for `memory leak` within tdlib repo and found [this](https://github.com/tdlib/td/issues/2116) issue. I think it may be related. 
Tldr; it's not a bug, it's a feature :)

Obviously it may not be the right answer. I'll check it by myself in a couple of days.

**Comment by anwsonwsymous** [comment]
Yeah I saw that and unfortunately this doesn't apply to our situation 😊
That is more about compilation

Maybe [this ](https://github.com/tdlib/td/issues/148#issuecomment-376942628) is the same issue ? 🤷‍♂️

They said that memory must be deallocated after 60 seconds if message database is used. But in my case deallocation is not happening for some reason.
Maybe we need drop received data manually? (I'm not good at unsafe rust)

**Comment by antonio-antuan** [comment]
Whoops, sorry, I do not know how it happened. 
[Here](https://github.com/tdlib/td/issues/1676) is the correct issue about (same?) memory leak and possible solutions.

**Comment by anwsonwsymous** [comment]
So currently, there is only one solution and it is [this](https://github.com/tdlib/td/issues/1676#issuecomment-926392375)

**Comment by antonio-antuan** [comment]
I'm not sure that restart the whole application is the best way to deal with it. Sounds weird, doesn't it? I need to think about the problem a bit.

I didn't have enough time for this but I hope that it will change in the coming days.

**Comment by antonio-antuan** [comment]
@anwsonwsymous While I don't yet have enough time for tests, you can test yourself with the following theory. If you want, of course :)

All related issues really say that there is only one way to deal with high memory consumption: you really need to close tdlib instance. [c-interface docs says](https://github.com/tdlib/td/blob/master/td/telegram/td_json_client.h#L33) that `TDLib client instances are destroyed automatically after they are close`. 

rust-tdlib client provides method `Client::close()` method. So it looks like you can try it. But, of course, after this you need to open instance again with the new `Worker::bind_client(...)` call.

**Comment by anwsonwsymous** [comment]
Yes of course, I will try it and let you know about the results here 👏

**Comment by anwsonwsymous** [comment]
Closing client helps!

Heaptrack result
![image](https://user-images.githubusercontent.com/30159288/190861885-1b451c56-b16f-42d1-9e2a-9656318d7690.png)


But now we/I have another problem. 
What if I want to use `ConsoleAuthStateHandler` only for the first time and after first reload always use `SignalAuthStateHandler` for next times. 
How can I do that without dynamic dispatch?

**Comment by antonio-antuan** [comment]
Currently there is no way to do it.
I'm going to implement `Worker::reload_client` method. I want it to close client (as you did it manually) and after this worker will try to re-authorize client. So, all you need is just call `reload_client`. 

Would it be suitable for you?

**Comment by anwsonwsymous** [comment]
Yes it will be perfect 👍🏼 
Can we also change `Worker` like this to have ability to change auth handler at runtime?

```
#[derive(Debug, Clone)]
pub struct Worker<S>
where
    S: TdLibClient + Send + Sync + Clone + 'static,
{
    run_flag: Arc<AtomicBool>,
    auth_state_handler: Arc<dyn AuthStateHander>,
    read_updates_timeout: Duration,
    channels_send_timeout: Duration,
    tdlib_client: S,
    clients: Arc<RwLock<ClientsMap<S>>>,
}
```

**Comment by makorne** [comment]
> Currently there is no way to do it. I'm going to implement `Worker::reload_client` method. I want it to close client (as you did it manually) and after this worker will try to re-authorize client. So, all you need is just call `reload_client`.
> 
> Would it be suitable for you?

Hi!
How about `Worker::reload_client? 
It would be great!

**Comment by antonio-antuan** [comment]
missed the ticket because it's closed, sorry.
reopened, going to implement it in a while.

**Comment by antonio-antuan** [comment]
resolved via #26 , published as 0.4.2
